### ISSUE TITLE
Unsoundness with bit-vectors.

### ISSUE THREAD

**Issue by hra687261** [OCamlPro/alt-ergo#664]
Given the following file: 
```
logic T: bitv[2]

goal g:
    ( T = [|00|] )
  or
    ( [|00|] <> ( [|0|] @ T^{0,0}) )
```
or
```
(set-logic QF_BV)
(declare-fun T () (_ BitVec 2))
(assert
  (and
    (distinct T (_ bv0 2))
    (= (_ bv0 2) (concat (_ bv0 1) ((_ extract 0 0) T)))))
(check-sat)
```
AE responds with `Valid` (`unsat`) while it's obvious that the goal is sat for `T = [|01|]` for example.

**Comment by bclement-ocp** [comment]
@hra687261 what commit did you test this on? I get `Fatal error: exception File "src/lib/reasoners/shostak.ml", line 506, characters 12-18: Assertion failed`

**Comment by hra687261** [comment]
Right, it requires the fix from #472, so either that PR or #669
I guess I'll reopen a PR with the fix in #472, because that one contains comments and modifications that are not part of the fix.

**Comment by bclement-ocp** [comment]
The issue is probably in the fix from #472 [here](https://github.com/OCamlPro/alt-ergo/pull/472/files#diff-f5c5d1437bc32718cda79735910071ec38afb11cd0701bae3498f276861d7a36R591-R593) but I am not entirely convinced the solution from #472 is correct, investigating a bit more
